cypress/e2e/shared/fluxQueryBuilder.test.ts

```diff
@@ -344,7 +345,99 @@ describe('Script Builder', () => {
         )
       })
 
-      it('should clear the editor text, and schema browser, with a new script', () => {
+      describe('conditions for divergence:', () => {
+        it('diverges when typing in composition block', () => {
+          cy.getByTestID('flux-sync--toggle').should('have.class', 'active')
+          cy.getByTestID('flux-editor', {timeout: 30000})
+          selectSchema()
+          confirmSchemaComposition()
+
+          cy.log('does not diverge when outside block')
+          cy.getByTestID('flux-editor').monacoType('// will not diverge')
+          cy.getByTestID('flux-sync--toggle').should(
+            'not.have.class',
+            'disabled'
+          )
+
+          cy.log('does diverge, within block')
+          cy.getByTestID('flux-editor').monacoType(
+            '{enter}{upArrow}{upArrow}{upArrow} // make diverge'
+          )
+          cy.log('toggle is now disabled')
+          cy.getByTestID('flux-sync--toggle').should('have.class', 'disabled')
+        })
+
+        describe('using hotkeys:', () => {
+          const runTest = (hotKeyCombo: string) => {
+            cy.getByTestID('flux-sync--toggle').should('have.class', 'active')
+            cy.getByTestID('flux-editor', {timeout: 30000})
+            selectSchema()
+            confirmSchemaComposition()
+
+            cy.log('does not diverge when outside block')
+            cy.getByTestID('flux-editor').monacoType(`foo ${hotKeyCombo}`)
+            cy.getByTestID('flux-sync--toggle').should(
+              'not.have.class',
+              'disabled'
+            )
+
+            cy.log('does diverge, within block')
+            cy.getByTestID('flux-editor').monacoType(
+              `{upArrow}{upArrow}{upArrow} ${hotKeyCombo}`
+            )
+            cy.log('toggle is now disabled')
+            cy.getByTestID('flux-sync--toggle').should('have.class', 'disabled')
+          }
+
+          it('diverges when commenting line', () => {
+            runTest('{cmd}/')
+          })
+
+          it('diverges when adding lines', () => {
+            runTest('{shift+alt+downArrow}')
+          })
+
+          it('diverges when removing lines', () => {
+            runTest('{cmd+x}')
+          })
+
+          it('diverges when moving lines', () => {
+            runTest('{alt+downArrow}')
+          })
```

Comment:
Each `it` block does run all of the beforeEach().

These test current run 2 beforeEaches:
* the login
* the clearing/reset of the script

The login can be switched to a before(), but the latter needs to stay a beforeEach()...since we need to reset the script and un-diverge (from the last divergence).

Pushing up the change now, to make the login a single before().

Reply:
Experiment confirmed. We do need the login as a beforeEach as well. Otherwise tests will fail. (See monitor-ci responds from single example commit.)

Will leave in both as a beforeEach for now. Total test time has been ~4 minutes (compared to ~8min flows test). Since the divergence feature should hopefully removed by end of Q4, then these tests will also be removed then.

Reply:
> Wait, don't we want it as a `beforeEach` to isolate each subsequent describe's tests?

Not sure I understand your question. Going to guess at what you meant...and please redirect if I'm missing your point. 😅 

Each test block is an `it` block. So in the case of this:
```
describe('outer scope/context', () => {
    beforeEach(() => {
        console.log('setup outer context')
    })
    describe('inner scope/context 1', () => {
        beforeEach(() => {
            console.log('setup inner context 1')
        })
        it('test A', () => {})
        it('test B', () => {})
    })
    describe('inner scope/context 2', () => {
        it('test C', () => {})
        it('test D', () => {})
    })
})
```

Test A & B will each get the outer and inner scopes run, per each test.
```
'setup outer context'
'setup inner context 1'
'test A'
'setup outer context'
'setup inner context 1'
'test B'
```

The positioning of the beforeEach, whether inside or outside a describe, is NOT about how often the beforeEach runs (since it runs per `it` block). Instead, it's about scoping the context setup.

If we want the "outer scope" beforeEach to run only once per each inner context --> then I believe we have to pull it into each inner scope describe block, and convert it to a before. So it'll be code duplication.

Please feel free to double check this. But it's what I understood from the docs, and then experimenting locally. 🤷🏼